### ISSUE TITLE
GfQuat types are missing GfIsClose

### ISSUE THREAD

**Issue by andrewkaufman** [PixarAnimationStudios/USD#3161]
### Description of Issue

As mentioned in #wg-usd on ASWF slack https://academysoftwarefdn.slack.com/archives/C013Z5AMT7T/p1720807979342119?thread_ts=1720804075.436449&cid=C013Z5AMT7T, the various GfQuat types do not work with `Gf.IsClose` in python. Presumbly they don't have a c++ implementation either.

**Comment by spiffmon** [comment]
That is correct.  Deciding what "the right" comparison should be for two quaternions is can be context-sensitive, I think.

**Comment by andrewkaufman** [comment]
Oh? Interesting. The context of the slack conversation is for a diffing tool. In that context, would an element wise comparision (as though it was a float4) not be correct?

**Comment by spiffmon** [comment]
For a straight-up textual or "numdiff" differ, yes that would be sufficient.  But if we provide a GfIsClose for quaternions, then someone might believe it's going to give them a reasonable answer about the angular distance between the two orientations, which a component-wise diff does not discriminate well.

**Comment by meshula** [comment]
I was thinking that a reasonable GfIsClose might be

abs(1 - abs(q1 x q2*)) < eps

where x is a multiplication and * is the inverse. That would capture the case where two rotations are the same but have an opposite sense.

Removing the inner abs would reject two rotations that are the same but have an opposite sense.

I wonder if GfIsClose should try to say that two quarternions are close with the same sense. And we could document how to accomplish the other.

**Comment by meshula** [comment]
Ps, I am suggesting the norm comparison rather than an element wise comparison because the elements of a quaternion are coupled, and the norm comparison will test the distribution of error about the sphere. A component wise comparison might report a lesser or greater difference but might represent a deviation from a norm based comparison. In other words the norm measure will capture a deviation from a unit length, and the component wise comparison will not.

**Comment by jesschimein** [comment]
Filed as internal issue #USD-9843

**Comment by meshula** [comment]
Here is suggestion on how to answer the GfIsClose question:


# Color

Color simply uses a V3 comparison.

```cpp
inline bool
GfIsClose(GfColor const &c1, GfColor const &c2, double tolerance)
{
    return GfIsClose(c1.GetRGB(), c2.GetRGB(), tolerance);
}
```

# Matrices

2d, 2f, 3d, 3f, 4d, 4f all follow this template:

```cpp
bool
GfIsClose(GfMatrix2d const &m1, GfMatrix2d const &m2, double tolerance)
{
    for(size_t row = 0; row < 2; ++row) {
        for(size_t col = 0; col < 2; ++col) {
            if(!GfIsClose(m1[row][col], m2[row][col], tolerance))
                return false;
        }
    }
    return true;
}
```

This is a simple value comparison. As the matrix has no interpretation imposed, it is impossible to create a meaningfully more precise comparison. For example, comparing two projection matrices could meaningfully have a better comparison. Comparing two regular TRS transformation matrices might meaningfully be better compared by decomposition and appropriate Euclidean comparisons.

# Vectors

Vector comparison follows this pattern, making an assumption that a Euclidean norm is intended for the vector. This is valid because all defined operations on the vector classes fundamentally impose a Euclidean interpretation.

```cpp
/// Tests for equality within a given tolerance, returning \c true if the
/// length of the difference vector is less than or equal to \p tolerance.
inline bool
GfIsClose(GfVec2f const &v1, GfVec2f const &v2, double tolerance)
{
    GfVec2f delta = v1 - v2;
    return delta.GetLengthSq() <= tolerance * tolerance;
}
```

# Quaternions

Quaternions declare that they are not specifically rotations:

```cpp
/// This class represents a generalized complex number that has a scalar real
/// part and a vector of three imaginary values. Quaternions are used by the
/// \c GfRotation class to represent arbitrary-axis rotations.
```

The Slerp function will not function meaningfully if the quaternion is not a unit quaternion, so it is an oddball. Slerping an aribtrary quaternion even after normalization is a bit odd, because not all quaternions represent S3 rotations.

```cpp
    /// Spherically interpolate between \p q0 and \p q1.
    ///
    /// If the interpolant \p alpha
    /// is zero, then the result is \p q0, while \p alpha of one yields
    /// \p q1.
    GF_API
    friend GfQuaternion GfSlerp(double alpha,
                                const GfQuaternion& q0,
                                const GfQuaternion& q1);
```                                

# Rotation
    
GfRotation is a similar to a quaternion class with rotation imposed but does not use GfQuaternion; however it can be constructed from quaternion and a quaternion may be derived from it.

```cpp
/// \class GfRotation
/// \ingroup group_gf_LinearAlgebra
///
/// Basic type: 3-space rotation specification.
///
/// This class represents a rotation in 3-space. This stores an axis as a
/// normalized vector of 3 \c doubles and an angle in degrees (as a double).
/// Rotations follow the right-hand rule: a positive rotation about an axis
/// vector appears counter-clockwise when looking from the end of the vector
/// toward the origin.
```

# Conclusion

Therefore, GfQuaternion should have a mathematically canonical comparison that is meaningful for quaternions, and GfRotation should have a GfClose that respects the rotational nature of the quaternion.

```diff
diff --git a/pxr/base/gf/quaternion.h b/pxr/base/gf/quaternion.h
index 06ec846f4..36829a1ad 100644
--- a/pxr/base/gf/quaternion.h
+++ b/pxr/base/gf/quaternion.h
@@ -231,6 +231,16 @@ GfDot(const GfQuaternion &q1, const GfQuaternion &q2) {
     return  (q1.GetReal() * q2.GetReal()) + GfDot(q1.GetImaginary(), q2.GetImaginary());
 }
 
+/// Tests for equality within a given tolerance, returning \c true if the
+/// Euclidean distance between the quaternions (as 4D vectors) is less than
+/// or equal to \p tolerance.
+inline bool
+GfIsClose(GfQuaternion const &q1, GfQuaternion const &q2, double tolerance)
+{
+    GfQuaternion delta = q1 - q2;
+    return GfDot(delta, delta) <= tolerance * tolerance;
+}
+
 PXR_NAMESPACE_CLOSE_SCOPE
 
 #endif // PXR_BASE_GF_QUATERNION_H
diff --git a/pxr/base/gf/rotation.h b/pxr/base/gf/rotation.h
index 1c1c2adef..5262a1562 100644
--- a/pxr/base/gf/rotation.h
+++ b/pxr/base/gf/rotation.h
@@ -274,6 +274,18 @@ class GfRotation {
 /// \ingroup group_gf_DebuggingOutput
 GF_API std::ostream& operator<<(std::ostream&, const GfRotation&);
 
+/// Tests for equality within a given tolerance, returning \c true if the
+/// cosine proximity on the unit quaternion sphere is within \p tolerance.
+/// A tolerance of 0 requires an exact match; ~0.001 corresponds to roughly
+/// 5 degrees of geodesic angular distance.
+inline bool
+GfIsClose(GfRotation const &r1, GfRotation const &r2, double tolerance)
+{
+    GfQuaternion q1 = r1.GetQuaternion();
+    GfQuaternion q2 = r2.GetQuaternion();
+    return (1.0 - std::fabs(GfDot(q1, q2))) <= tolerance;
+}
+
 PXR_NAMESPACE_CLOSE_SCOPE
 
 #endif // PXR_BASE_GF_ROTATION_H
```

**Comment by nporcino-pixar** [comment]
Chatted with @nvmkuruc and do agree with him that even though component wise comparison of a Quaternion is not mathematically meaningful, it is least surprising for unit tests where we are checking for specific numeric values as results as opposed to mathematically meaningful "close" values. If we prefer the unit-test meaning of close, I'd prefer that we document Quaternion's GfIsClose in the header as being the wrong thing to reach for if you want to know if Quaternions are close in a mathematical sense. That's true for all the GfIsClose methods, and the fact that GfVec is a squared distance Euclidean test definitely confuses unit test concerns (are the numerical values roughly equal) with mathematical concerns (are two values close in a Euclidean sense?), prompting this whole debate.

If the idea is that GfIsClose is strictly for unit test comparisons of numeric values, and not mathematical closeness, then the the GfRotation should also be element-wise, and similarly, I think we should document that GfIsClose is not useful for math, only test comparisons.
